src/allmydata/test/test_i2p_provider.py

```diff
@@ -294,6 +294,28 @@ def test_handler_default(self):
         self.assertIs(h, handler)
         i2p.default.assert_called_with(reactor, keyfile=None)
 
+class Provider_Listener(unittest.TestCase):
+    def test_listener(self):
+        i2p = mock.Mock()
+        handler = object()
+        i2p.local_i2p = mock.Mock(return_value=handler)
+        reactor = object()
+
+        privkeyfile = os.path.join("private", "i2p_dest.privkey")
+        with mock_i2p(i2p):
+            p = i2p_provider.Provider("basedir",
+                                      FakeConfig(**{
+                                          "i2p.configdir": "configdir",
+                                          "sam.port": "goodport",
+                                          "dest": "true",
+                                          "dest.port": "3457",
+                                          "dest.private_key_file": privkeyfile,
+                                          }),
+                                      reactor)
+            endpoint_or_description = p.get_listener()
+        self.assertEqual(endpoint_or_description,
+                         "i2p:%s:3457:api=SAM:apiEndpoint=goodport" % privkeyfile)
```

Comment:
Naw, the whole point of this exercise is to avoid the need for escaping anything. In this early state, the `get_listener()` method doesn't return anything complicated enough to require escaping. In the later state (which I hope @str4d will write), it will return an Endpoint object instead of a (string) endpoint descriptor, so it won't need escaping either.